File: DataAlgo_forPy/Search_Sort/selection_sort.rb
# Implement selection sort

def selection_sort(array)
  reference_index = 0

  until reference_index == array.size - 1
    minimum = array[reference_index]
    minimum_index = reference_index

    (reference_index...array.size).each do |i|
      if array[i] < minimum
        minimum = array[i]
        minimum_index = i
      end
    end

    if minimum_index != reference_index
      array[reference_index], array[minimum_index] = array[minimum_index], array[reference_index]
    end

    reference_index += 1
  end

  array
end

p selection_sort([5,4,3,2,1]) # [1,2,3,4,5]

File: DataAlgo_forPy/Search_Sort/selection_sort.js
// Implement selection sort

const selection_sort = (array) => {
  let reference_index = 0

  while (reference_index < array.length - 1) {
    let minimum = array[reference_index]
    let minimum_index = reference_index

    for (var i = reference_index; i < array.length; i++) {
      if (array[i] < minimum) {
        minimum = array[i]
        minimum_index = i
      }
    }

    if (minimum_index !== reference_index) {
      let temp = array[reference_index]
      array[reference_index] = array[minimum_index]
      array[minimum_index] = temp
    }

    reference_index++
  }

  return array
}

console.log(selection_sort([5,4,3,2,1])) // [1,2,3,4,5]
